tutorials/flappy-bird/step_8/flappy-bird.cpp

```diff
@@ -0,0 +1,660 @@
+            new_rotation = constants.max_angle;
+
+        // Set the properties
+        registry.assign_or_replace<transform::properties>(player, transform::properties{.rotation = new_rotation});
+    }
+
+private:
+    entt::entity player;
+    math::vec2f movement_speed{0.f, 0.f};
+};
+
+// Name this system
+REFL_AUTO (type(player_logic));
+
+// Collision Logic System
+class collision_logic final : public ecs::logic_update_system<collision_logic> {
+public:
+    collision_logic(entt::registry &registry, entt::entity player_, bool &player_died_) noexcept : system(registry),
+                                                                                                   player(player_),
+                                                                                                   player_died(player_died_) {}
+    // Update, this will be called every tick
+    void update() noexcept final {
+        auto &registry = entity_registry_;
+
+        // Do not check anything if player is already dead
+        if (player_died) return;
+
+        // Loop all columns to check collisions between player and the pipes
+        for (auto entity : registry.view<graphics::layer<3>>()) {
+            // Check collision between player and a collidable object
+            if (collisions::basic_collision_system::query_rect(registry, player, entity)) {
+                // Mark player died as true
+                player_died = true;
+            }
+        }
+    }
+
+private:
+    entt::entity player;
+    bool &player_died;
+};
+
+// Name this system
+REFL_AUTO (type(collision_logic));
+
+// Game Scene
+class game_scene final : public scenes::base_scene {
+public:
+    game_scene(entt::registry &registry, ecs::system_manager &system_manager_) noexcept : base_scene(registry),
+                                                                                          system_manager(
+                                                                                                  system_manager_) {
+        // Set the constants that will be used in the program
+        registry.set<flappy_bird_constants>();
+
+        // Create everything
+        score_entity = create_score(registry);
+        create_background(registry);
+        init_dynamic_objects(registry);
+    }
+
+    // Scene name
+    std::string scene_name() noexcept final {
+        return "game_scene";
+    }
+
+private:
+    // Update the game every tick
+    void update() noexcept final {
+        // Retrieve constants
+        const auto constants = entity_registry_.ctx<flappy_bird_constants>();
+
+        // Check if jump key is tapped
+        bool jump_key_tapped = input::virtual_input::is_tapped("jump");
+
+        // If game is not started yet and jump key is tapped
+        if (!started_playing && jump_key_tapped) {
+            // Game starts, player started playing
+            started_playing = true;
+            resume_physics();
+        }
```

Comment:
checking a state again here, 3 separate function for each step of the update function with a good naming will be more explicit